src/cdomains/intDomain.ml

```diff
@@ -1861,193 +1861,247 @@ module Booleans = MakeBooleans (
+  end
+  type t = Inc of ISet.t | Exc of ISet.t * R.t (* inclusion/exclusion set *)
+
+  type int_t = BI.t
   let name () = "enums"
-  let top_range = R.of_interval range_ikind (-99L, 99L) (* Since there is no top ikind we use a range that includes both ILongLong [-63,63] and IULongLong [0,64]. Only needed for intermediate range computation on longs. Correct range is set by cast. *)
-  let bot () = Inc []
-  let top_of ik = Exc ([], size ik)
-  let top () = Exc ([], top_range)
-  let top_of ik = top ()
-  let bot_of ik = bot ()
-  let is_top x = x = top ()
+  let to_yojson x = failwith @@ "to_yojson unimplemented for " ^ (name ())
+
+  let bot () = failwith "bot () not implemented for Enums"
+  let top_of ik = Exc (ISet.empty, size ik)
+  let top () = failwith "top () not implemented for Enums"
+  let bot_of ik = Inc ISet.empty
 
-  let equal a b = a = b (* Be careful: this works only as long as the Range/Interval implementation used does not use big integers *)
+  let min_int ik = I.of_bigint @@ fst @@ Size.range_big_int ik
+  let max_int ik = I.of_bigint @@ snd @@ Size.range_big_int ik
+
+  let equal u v = match u, v with
+    | Inc x, Inc y -> ISet.equal x y
+    | Exc (x, r), Exc (y, s) -> ISet.equal x y && R.equal r s
+    | _, _ -> false
   let short _ = function
-    | Inc[] -> "bot" | Exc([],r) -> "top"
-    | Inc xs -> "{" ^ (String.concat ", " (List.map (I.short 30) xs)) ^ "}"
-    | Exc (xs,r) -> "not {" ^ (String.concat ", " (List.map (I.short 30) xs)) ^ "} " ^ "("^R.short 2 r^")"
+    | Inc xs when ISet.is_empty xs -> "bot"
+    | Inc xs -> "{" ^ (String.concat ", " (List.map (I.short 30) (ISet.to_list  xs))) ^ "}"
+    | Exc (xs,r) -> "not {" ^ (String.concat ", " (List.map (I.short 30) (ISet.to_list xs))) ^ "} " ^ "("^R.short 2 r^")"
+
   include Std (struct type nonrec t = t let name = name let top_of = top_of let bot_of = bot_of let short = short let equal = equal end)
 
+  let compare a b =
+    let value c =
+      match c with Inc _ -> 0 | Exc _ -> 1
+    in
+    match a, b with
+    | Inc x, Inc y -> ISet.compare x y
+    | Exc (x,r), Exc (y, s) ->
+        let comp_sets = ISet.compare x y in
+        if comp_sets = 0 then R.compare r s else comp_sets
+    | _, _ -> compare (value a) (value b)
+
+  let hash = function
+    | Inc x ->
+      let m = match BI.to_int (ISet.min_elt x) with n -> n | exception Not_found -> 41 | exception Failure _ -> 61 in (* Failure occurs when the conversion to int did not work, Not_found occurs when the set is empty *)
+      let n = match BI.to_int (ISet.max_elt x) with n -> n | exception Not_found -> 43 | exception Failure _ -> 67 in
+      3  * ISet.cardinal x + 7 * m + 13 * n
+    | Exc (x, r) ->
+      let m = match BI.to_int (ISet.min_elt x) with n -> n | exception Not_found  -> 53 | exception Failure _ -> 71 in
+      let n = match BI.to_int (ISet.max_elt x) with n -> n | exception Not_found  -> 59 | exception Failure _ -> 73 in
+      R.hash r + 5  * ISet.cardinal x + 11 * m + 19 * n
```

Comment:
Is there a reason why this is so ad-hoc and involved? If it's because `ISet.hash` doesn't exist, one could just use `SetDomain.Make` instead, like `DefExc`.

Reply:
Thanks, I am now using `SetDomain.Make`.